# Benchmark Results Analysis

## Environment

This project was executed on a Ubuntu 22.04 container with the following specs:

- Host
    - Memory:
        - Capacity: 16GB
        - Speed: 3200Mhz
        - SWAP: 2GB
    - CPU:
        - Model: i5-1135G7
        - BaseClock: 2.4Ghz
        - Cores: 4 Cores and 8 Threads
        - Turbo: 4.2Ghz
        - L1: 128Kb
        - L2: 5Mb
        - L3: 8Mb
    - Disk:
        - Speed: 2.3Gb/s
        - Capacity: 500GB
    - System:
        - Distro: Ubuntu 22.04
- Container:
    - Memory:
        - Capacity: 6GB
        - SWAP: 2GB
    - CPU:
        - Cores: 4 Threads
    - System:
        - Distro: Ubuntu 22.04

## Keras NN vs Numpy BNN

### Data Explanation

In [1]:
import pandas as pd

In [2]:
results = pd.read_csv("data/benchmark.csv")
results.columns = results.columns.str.replace(r"\\", "", regex=False)
results

name        case   time.min   time.max  time.median  time.mean  \
0          mnist         bit   0.000729   0.012377     0.000918   0.000990   
1          mnist       keras   0.033221   0.123128     0.044749   0.047120   
2          mnist  comparison  45.584796   9.948382    48.751029  47.602551   
3        housing         bit   0.001180   0.003951     0.001412   0.001521   
4        housing       keras   0.034954   0.117986     0.040935   0.042585   
5        housing  comparison  29.626293  29.865567    28.987917  28.006006   
6         iris-8         bit   0.001493   0.018302     0.001763   0.001972   
7         iris-8       keras   0.034915   0.175458     0.041313   0.043196   
8         iris-8  comparison  23.387012   9.586824    23.434564  21.904484   
9         iris-4         bit   0.001475   0.013956     0.001738   0.001873   
10        iris-4       keras   0.035059   0.147780     0.040703   0.042372   
11        iris-4  comparison  23.771355  10.589291    23.424221  22.619979   
12     mnist-128         bit   0.000831   0.011935     0.000985   0.001102   
13     mnist-128       keras   0.035106   0.135428     0.040550   0.042260   
14     mnist-128  comparison  42.262398  11.346831    41.184394  38.331303   
15   many-inputs         bit   0.000477   0.003376     0.000549   0.000598   
16   many-inputs       keras   0.034955   0.121691     0.040206   0.042086   
17   many-inputs  comparison  73.239630  36.048257    73.291715  70.368705   
18   many-layers         bit   0.001797   0.016474     0.002091   0.002258   
19   many-layers       keras   0.034957   0.145795     0.042447   0.044916   
20   many-layers  comparison  19.457986   8.849867    20.295817  19.889116   
21    many-units         bit   0.000911   0.005040     0.001094   0.001202   
22    many-units       keras   0.035167   0.126323     0.041572   0.043069   
23    many-units  comparison  38.618702  25.064123    37.998740  35.834473   
24  all-together         bit   0.002356   0.006880     0.002770   0.002982   
25  all-together       keras   0.035808   0.160201     0.042648   0.044284   
26  all-together  comparison  15.196466  23.286290    15.394573  14.852649   

     time.std time.unit  memory.value memory.unit  parameters_memory.value  \
0    0.000355   seconds     30.492188   kilobytes                53.234375   
1    0.009216   seconds     88.515625   kilobytes              1590.609375   
2   25.976789   seconds      2.902895   kilobytes                29.879366   
3    0.000304   seconds     39.531250   kilobytes                 3.593750   
4    0.007145   seconds    101.625000   kilobytes                46.156250   
5   23.504357   seconds      2.570751   kilobytes                12.843478   
6    0.000768   seconds     48.250000   kilobytes                 1.289062   
7    0.008345   seconds    115.539062   kilobytes                 1.898438   
8   10.863874   seconds      2.394592   kilobytes                 1.472727   
9    0.000562   seconds     48.031250   kilobytes                 1.195312   
10   0.007290   seconds    115.531250   kilobytes                 1.367188   
11  12.975893   seconds      2.405335   kilobytes                 1.143791   
12   0.000502   seconds     30.492188   kilobytes                13.765625   
13   0.007302   seconds     88.351562   kilobytes               398.109375   
14  14.559084   seconds      2.897515   kilobytes                28.920545   
15   0.000167   seconds     20.687500   kilobytes                 4.460938   
16   0.007496   seconds     74.929688   kilobytes               128.460938   
17  44.967971   seconds      3.621979   kilobytes                28.796848   
18   0.000606   seconds     56.921875   kilobytes                 2.617188   
19   0.009835   seconds    129.429688   kilobytes                19.242188   
20  16.240625   seconds      2.273813   kilobytes                 7.352239   
21   0.000328   seconds     30.492188   kilobytes               140.570312   
22   0.007268   seconds     88.44

The table above shows the time and memory consumption of Keras model created from dense layers and an authoral implementation of the bit neural network using numpy benchmarked 2000 times.
This benchmark was executed only on the feed-forward step of the network.

Names refeer to different defined topologies which are executed both in the bit and keras implementation, while the comparison is just the proportion of the metrics of keras to bit.

In [3]:
import yaml
import pprint
with open("../config/benchmark.yaml") as f:
    topologies = yaml.load(f, yaml.loader.FullLoader)
pprint.pprint(topologies["experiment"]["cases"])

{'all-together': {'inputs': 1024, 'units': [1024, 1024, 1024, 1024, 1024]},
 'housing': {'inputs': 13, 'units': [100, 100, 1]},
 'iris-4': {'inputs': 4, 'units': [4, 4, 4, 3]},
 'iris-8': {'inputs': 4, 'units': [8, 8, 8, 3]},
 'many-inputs': {'inputs': 1024, 'units': [32]},
 'many-layers': {'inputs': 10, 'units': [32, 32, 32, 32, 32]},
 'many-units': {'inputs': 10, 'units': [1024, 1024]},
 'mnist': {'inputs': 784, 'units': [512, 10]},
 'mnist-128': {'inputs': 784, 'units': [128, 10]}}


### Methodology

If we look at the first table we are going to see sometimes the max and min values are too far from the `median`, this means that the `mean` is not a good metric since it will consider these outlier. In reason of that, we are going to use only `median` for comparing time.

In [4]:
{
    "max": (results["time.median"] / results["time.max"]).mean(),
    "min": (results["time.median"] / results["time.min"]).mean(),
}

{'max': 0.9274996622624422, 'min': 1.1295207481333718}

Another thing we can see is that the memory consumption is very different from the `parameters_memory` column. This happened because the memory value was measuring the size of the network `object` while the `parameters_memory` is the real size occupied by the network. In reason of that, we are going to consider only the `parameter_memory`.

In [5]:
results = results[["name", "case", "time.median", "time.std", "parameters_memory.value"]]

## Analysis

In [6]:
results[results["case"] == "comparison"]

name        case  time.median   time.std  parameters_memory.value
2          mnist  comparison    48.751029  25.976789                29.879366
5        housing  comparison    28.987917  23.504357                12.843478
8         iris-8  comparison    23.434564  10.863874                 1.472727
11        iris-4  comparison    23.424221  12.975893                 1.143791
14     mnist-128  comparison    41.184394  14.559084                28.920545
17   many-inputs  comparison    73.291715  44.967971                28.796848
20   many-layers  comparison    20.295817  16.240625                 7.352239
23    many-units  comparison    37.998740  22.135037                29.483966
26  all-together  comparison    15.394573  12.956367                30.998464

The comparison table above shows the **speedup** in the time columns while the `parameters_memory` shows the **size compression rate** of the bit compared to the keras

### Time

If we look at the comparison table, we are going to see the behaviour of the time and the memory is not proportional. This happens due to the real reason of the speedup.

In [7]:
results[results["case"] == "keras"][["name", "case", "time.median"]]

name   case  time.median
1          mnist  keras     0.044749
4        housing  keras     0.040935
7         iris-8  keras     0.041313
10        iris-4  keras     0.040703
13     mnist-128  keras     0.040550
16   many-inputs  keras     0.040206
19   many-layers  keras     0.042447
22    many-units  keras     0.041572
25  all-together  keras     0.042648

By looking at the table above we are going to see the times do not vary according to the size of the network. For instance, take the iris-4 and the mnist-128. The iris 4 has `13*4 + 4*4 + 4*4 + 4*3 = 96` parameters while the mnist-128 contains`784*128 + 128*10 = 101632` parameters.

This happens because Keras backend (Tensorflow) performs a lot of optimizations on the math operations. Some of the observed ones were:

- Multithreading: If we look at the CPU usage while executing Keras `predict` method we are going to see more than one core is being used at a time
- SIMD: When importing Keras or Tensorflow we are going to see a message telling us the binary uses SSE and AVX to vectorize operations
- Memory Manipulation: because the Tensorflow operations are implemented in low-level APIs the code is written focusing on the best memory usage, avoiding useless allocations, and cache optimization
- Loop: Similarly to the previous, because it is implemented in low-level API every loop operation is only executed once. Differently from the high-level APIs such as Numpy, where every operation is containerized in its own scope, so if you need to perform multiple transformations, it will have to loop through various method containers
- BLAS: Tensorflow uses long-term developed arithmetic libraries such as BLAS, iMKL (Intel), ATLAS (ARM), BLIS (AMD) for performing its operations which makes them much more efficient in terms of execution time
- Cycles Per Instruction: Number of CPU cycles required to execute an instruction can influence the code performance depending on the hardware executing it

All these factors lead us to know the real bottleneck on the Keras code is preparing the input, output, validation or any other internal logic and the calculation itself.

In [46]:
import numpy as np
import timeit
a = np.zeros((1000,), dtype=np.uint32)
b = np.ones((1000, 1000), dtype=np.uint32)
time = lambda fn: print("Spent: %f" % (timeit.timeit(fn, number=1000) / 1000))
time(lambda: np.dot(a, b))
time(lambda: np.bitwise_xor(a, b))

Spent: 0.000839
Spent: 0.000366


Take a look at the time spent above. XOR is only one of the operations executed by the bit network feed-forward and it takes approximately half the time as the dot product in numpy which is most of the compute intensive work required to run feed-forward on a traditional network.

In [19]:
keras_bit = keras_bit[["name", "case", "time.median", "time.std"]]
keras_bit = results[results["case"].isin(["keras", "bit"])]
keras_bit = keras_bit.assign(pct_std=keras_bit["time.std"] / keras_bit["time.median"])
keras_bit.groupby("case")["pct_std"].mean()

case
bit      0.331002
keras    0.190646
Name: pct_std, dtype: float64

Another interesting thing to see is the standard deviation in time consumption of each network type. If we look at the bit network we are going to see the time spent by the bit network varies in 33% from them median while the keras network varies 19%. This may indicate a better memory usage once the memory access time is not constant.

### Memory

On the other hand, when we look at the memory comparison we are going to see a huge difference proportional to the size of the network. Because we used an uin32 as the weights container, the maximum improvement in compression is close to 32x.

In [48]:
results[results["case"] == "comparison"]

name        case  time.median   time.std  parameters_memory.value
2          mnist  comparison    48.751029  25.976789                29.879366
5        housing  comparison    28.987917  23.504357                12.843478
8         iris-8  comparison    23.434564  10.863874                 1.472727
11        iris-4  comparison    23.424221  12.975893                 1.143791
14     mnist-128  comparison    41.184394  14.559084                28.920545
17   many-inputs  comparison    73.291715  44.967971                28.796848
20   many-layers  comparison    20.295817  16.240625                 7.352239
23    many-units  comparison    37.998740  22.135037                29.483966
26  all-together  comparison    15.394573  12.956367                30.998464

You may be wondering why smaller networks have a compression rate smaller than the larger ones. That can be explained by the wasted bits. For instance, let's consider a network that has 1 input and a layer with 100 units, this would make a layer with a weight matrix of size 100x1. Since we are using an uint32, i.e. 32 bits, to compress our weights, for each unit we are going to use 1 bit of the uint32 weight, thus in this specific network there's no compression at all because the weights matrix would have 100x1 size.

For smaller networks, using an uint8 or an uint16 would be a better idea if memory is a bottleneck, otherwise it won't be something you need to bother. In larger networks, the bit waste problem is not even noticeable because you'll only lose at most 30 bits per unit.

Another factor that doesn't allow us to have a 32x compression is when we have a bias vector on our network. That happens because the bias vector may not be binarized as well. In the networks presented above this is the case.

## CPP Float NN vs CPP BNN

### Data Explanation

In [4]:
cpp = pd.read_csv("./data/benchmark_cpp.csv")
cpp.columns = cpp.columns.str.replace(r"\\", "", regex=False)
cpp

name        case   time.min   time.max  time.median  time.mean  \
0          mnist         bit   0.000052   0.000315     0.000059   0.000064   
1          mnist       float   0.001072   0.008807     0.001203   0.001246   
2          mnist  comparison  20.615385  27.958730    20.389831  19.549087   
3        housing         bit   0.000004   0.000109     0.000006   0.000006   
4        housing       float   0.000029   0.000152     0.000034   0.000036   
5        housing  comparison   7.250000   1.394495     5.666667   5.658249   
6         iris-8         bit   0.000001   0.000014     0.000002   0.000002   
7         iris-8       float   0.000001   0.000100     0.000002   0.000002   
8         iris-8  comparison   1.000000   7.142857     1.000000   0.889822   
9         iris-4         bit   0.000001   0.001950     0.000002   0.000003   
10        iris-4       float   0.000001   0.000021     0.000001   0.000001   
11        iris-4  comparison   1.000000   0.010769     0.500000   0.520253   
12     mnist-128         bit   0.000016   0.000119     0.000017   0.000018   
13     mnist-128       float   0.000295   0.002347     0.000321   0.000331   
14     mnist-128  comparison  18.437500  19.722689    18.882353  17.916286   
15   many-inputs         bit   0.000005   0.000095     0.000006   0.000007   
16   many-inputs       float   0.000097   0.001817     0.000102   0.000107   
17   many-inputs  comparison  19.400000  19.126316    17.000000  15.628049   
18   many-layers         bit   0.000003   0.000109     0.000004   0.000004   
19   many-layers       float   0.000011   0.000114     0.000012   0.000013   
20   many-layers  comparison   3.666667   1.045872     3.000000   2.938221   
21    many-units         bit   0.000160   0.007125     0.000170   0.000191   
22    many-units       float   0.003095   0.009804     0.003237   0.003322   
23    many-units  comparison  19.343750   1.376000    19.044118  17.438622   
24  all-together         bit   0.000705   0.006747     0.000768   0.000830   
25  all-together       float   0.014029   0.032940     0.015275   0.015685   
26  all-together  comparison  19.899291   4.882170    19.889974  18.887116   

        time.std time.unit  benchmark.repeat  
0   1.394052e-05   seconds              2000  
1   3.027962e-04   seconds              2000  
2   2.172058e+01   seconds              2000  
3   3.695287e-06   seconds              2000  
4   7.130329e-06   seconds              2000  
5   1.929574e+00   seconds              2000  
6   6.325433e-07   seconds              2000  
7   2.341802e-06   seconds              2000  
8   3.702201e+00   seconds              2000  
9   4.356069e-05   seconds              2000  
10  8.609400e-07   seconds              2000  
11  1.976415e-02   seconds              2000  
12  4.991250e-06   seconds              2000  
13  6.830541e-05   seconds              2000  
14  1.368503e+01   seconds              2000  
15  2.858090e-06   seconds              2000  
16  3.961363e-05   seconds              2000  
17  1.386018e+01   seconds              2000  
18  2.783386e-06   seconds              2000  
19  3.447720e-06   seconds              2000  
20  1.238678e+00   seconds              2000  
21  1.862589e-04   seconds              2000  
22  3.323950e-04   seconds              2000  
23  1.784586e+00   seconds              2000  
24  2.709691e-04   seconds              2000  
25  1.411802e-03   seconds              2000  
26  5.210195e+00   seconds              2000

This time we are comparing two authoral implementations made in C++ for benchmarking. The bit code is the bit neural network, while the float is the regular MLP implementation using float weights. The comparison is the proportion of float network stats to the bit network.

The topologies used in this benchmark are the same used in the previous benchmark.

### Methodology

The same considerations performed in the previous section will be replicated here.

## Analysis

### Time

In [5]:
cpp[cpp["case"] == "comparison"][["name", "case", "time.median"]]

name        case  time.median
2          mnist  comparison    20.389831
5        housing  comparison     5.666667
8         iris-8  comparison     1.000000
11        iris-4  comparison     0.500000
14     mnist-128  comparison    18.882353
17   many-inputs  comparison    17.000000
20   many-layers  comparison     3.000000
23    many-units  comparison    19.044118
26  all-together  comparison    19.889974

Looking at the table above we can see the time is proportional to the size of the network. This is what we expect since we have a great reduction in the number of operations performed due to the SIMD nature of the bit neural network.

If we look at the smaller networks we are going to see the less number of parameters, the lower the time difference. This happens because the other operations required for the bit dot product do not compensate the time used to multiply and sum the elements in a traditional network. Thus, the problem is a side-product of the bit waste effect mentioned before.

### Memory

Memory was not evaluated here because the number of parameters is the same of the previous experiment. Thus do not make difference.

## Optimized CPP Float NN vs CPP BNN

### Data Explanation

In [8]:
opt = pd.read_csv("./data/benchmark_cpp_opt.csv")
opt.columns = opt.columns.str.replace(r"\\", "", regex=False)
opt

name        case   time.min   time.max  time.median  time.mean  \
0          mnist         bit   0.000009   0.000123     0.000011   0.000012   
1          mnist       float   0.000560   0.009220     0.000640   0.000667   
2          mnist  comparison  62.222222  74.959350    58.181818  57.086895   
3        housing         bit   0.000001   0.000102     0.000002   0.000002   
4        housing       float   0.000011   0.000101     0.000013   0.000014   
5        housing  comparison  11.000000   0.990196     6.500000   7.188883   
6         iris-8         bit   0.000000   0.000024     0.000001   0.000001   
7         iris-8       float   0.000000   0.000032     0.000001   0.000001   
8         iris-8  comparison        NaN   1.333333     1.000000   0.989076   
9         iris-4         bit   0.000000   0.000008     0.000001   0.000001   
10        iris-4       float   0.000000   0.000010     0.000001   0.000001   
11        iris-4  comparison        NaN   1.250000     1.000000   0.991278   
12     mnist-128         bit   0.000003   0.000030     0.000004   0.000004   
13     mnist-128       float   0.000140   0.000492     0.000156   0.000163   
14     mnist-128  comparison  46.666667  16.400000    39.000000  38.252207   
15   many-inputs         bit   0.000001   0.000025     0.000002   0.000002   
16   many-inputs       float   0.000047   0.000125     0.000053   0.000055   
17   many-inputs  comparison  47.000000   5.000000    26.500000  27.765259   
18   many-layers         bit   0.000001   0.000047     0.000002   0.000002   
19   many-layers       float   0.000002   0.000087     0.000003   0.000003   
20   many-layers  comparison   2.000000   1.851064     1.500000   1.847567   
21    many-units         bit   0.000025   0.000157     0.000029   0.000031   
22    many-units       float   0.001520   0.007061     0.001708   0.001738   
23    many-units  comparison  60.800000  44.974522    58.896552  56.742338   
24  all-together         bit   0.000107   0.000413     0.000121   0.000128   
25  all-together       float   0.007628   0.017465     0.008416   0.008561   
26  all-together  comparison  71.289720  42.288136    69.553719  66.907282   

        time.std time.unit  benchmark.repeat  
0   4.602197e-06   seconds              2000  
1   2.680090e-04   seconds              2000  
2   5.823500e+01   seconds              2000  
3   2.388848e-06   seconds              2000  
4   2.955354e-06   seconds              2000  
5   1.237146e+00   seconds              2000  
6   7.822404e-07   seconds              2000  
7   8.268440e-07   seconds              2000  
8   1.057020e+00   seconds              2000  
9   4.393606e-07   seconds              2000  
10  4.968579e-07   seconds              2000  
11  1.130866e+00   seconds              2000  
12  1.468587e-06   seconds              2000  
13  1.889243e-05   seconds              2000  
14  1.286436e+01   seconds              2000  
15  9.411929e-07   seconds              2000  
16  5.489492e-06   seconds              2000  
17  5.832483e+00   seconds              2000  
18  1.339096e-06   seconds              2000  
19  2.046137e-06   seconds              2000  
20  1.527999e+00   seconds              2000  
21  7.175739e-06   seconds              2000  
22  1.960240e-04   seconds              2000  
23  2.731760e+01   seconds              2000  
24  1.898095e-05   seconds              2000  
25  7.683060e-04   seconds              2000  
26  4.047773e+01   seconds              2000

The data presented in this section refeers to the same C++ code, but this time compiled with the flag `-O3` and `-march=tigerlake` which means the compiler will perform stronger optimizations such as vectorization better memory management and build code focused on the specific hardware architecture.

### Time

In [10]:
opt[opt["case"] == "comparison"][["name", "case", "time.median"]]

name        case  time.median
2          mnist  comparison    58.181818
5        housing  comparison     6.500000
8         iris-8  comparison     1.000000
11        iris-4  comparison     1.000000
14     mnist-128  comparison    39.000000
17   many-inputs  comparison    26.500000
20   many-layers  comparison     1.500000
23    many-units  comparison    58.896552
26  all-together  comparison    69.553719

Looking at the comparison above we can see the larger networks are most likely to have the larger speedups. The problem with smaller networks is the same as mentioned before. However, in this optimized code we can see many units is the major reason for great speedups, and when we combine it with many inputs the speedup becomes even higher. This means it is not the number of inputs or the number of units that affect the speedup, but the number of parameters, i.e., the more parameters the more operations we would need to perform. 

### Sklearn MLP

### Data Explanation

In [11]:
sklearn = pd.read_csv("./data/benchmark_sklearn.csv")
sklearn.columns = sklearn.columns.str.replace("\\", "", regex=False)
sklearn

name  time.min  time.max  time.median  time.mean  time.std  \
0         mnist  0.000113  0.013749     0.000162   0.000231  0.000611   
1       housing  0.000067  0.008214     0.000099   0.000124  0.000286   
2        iris-8  0.000060  0.007011     0.000099   0.000100  0.000156   
3        iris-4  0.000060  0.000373     0.000066   0.000077  0.000020   
4     mnist-128  0.000072  0.008321     0.000079   0.000121  0.000319   
5   many-inputs  0.000065  0.007593     0.000094   0.000115  0.000284   
6   many-layers  0.000064  0.012837     0.000106   0.000121  0.000389   
7    many-units  0.000231  0.009055     0.000282   0.000333  0.000343   
8  all-together  0.001203  0.046884     0.001402   0.001587  0.001419   

  time.unit   memory.value memory.unit  parameters_memory.value  \
0   seconds    9547.843750   kilobytes              3180.984375   
1   seconds     279.976562   kilobytes                91.718750   
2   seconds      14.000000   kilobytes                 3.000000   
3   seconds      10.781250   kilobytes                 1.937500   
4   seconds    2392.843750   kilobytes               795.984375   
5   seconds     776.101562   kilobytes               257.078125   
6   seconds     117.632812   kilobytes                37.578125   
7   seconds   24895.343750   kilobytes              8296.828125   
8   seconds  123033.820312   kilobytes             41009.578125   

  parameters_memory.unit  benchmark.repeat  
0              kilobytes              2000  
1              kilobytes              2000  
2              kilobytes              2000  
3              kilobytes              2000  
4              kilobytes              2000  
5              kilobytes              2000  
6              kilobytes              2000  
7              kilobytes              2000  
8              kilobytes              2000

In this dataset we only benchmarked sklearn implementation of a regular MLP. This benchmark was performed because most of its code is written over numpy operations and do not require that much preparation steps like Keras. Thus we can use it to compare with our implementation without worrying too much about the framework internal logic.

### Time

In [20]:
bit = opt[opt["case"] == "bit"][["name", "time.median"]].rename(columns={"time.median": "time.median.bit"})
sklearn_bit = sklearn[["name", "time.median"]].rename(columns={"time.median": "time.median.skl"}).merge(bit, on="name")
sklearn_bit["time.median.comparison"] = sklearn_bit["time.median.skl"] / sklearn_bit["time.median.bit"]
sklearn_bit

name  time.median.skl  time.median.bit  time.median.comparison
0         mnist         0.000162         0.000011               14.744863
1       housing         0.000099         0.000002               49.629250
2        iris-8         0.000099         0.000001               99.133500
3        iris-4         0.000066         0.000001               66.219498
4     mnist-128         0.000079         0.000004               19.874376
5   many-inputs         0.000094         0.000002               47.195499
6   many-layers         0.000106         0.000002               52.931500
7    many-units         0.000282         0.000029                9.711690
8  all-together         0.001402         0.000121               11.582694

In this case we can see a speedup comparing the bit implementation with the sklearn one. However it is not as significant as the last example. This can be explained by looking at the smaller examples. Take iris-4 and many-inputs for instance, we can see sklearn speedup is not proportional in to the bit waste problem or to the network size. This is exaplained by the data validation or inner logic performed by the framework. Although sklearn has less pre-predict logic it is still there, and consumes time proportional to the size of the input. Another fact that indicate this same reason is looking at the all-together and to the many-units cases because they show us when computation is the most part of the prediction the speedup is not as high as when we have many inputs.

The reason why we don't have real speedups greater than 10x is due to the same reasons mentioned for Keras.